server/src/main/java/org/opensearch/cluster/service/Throttler.java

```diff
@@ -0,0 +1,93 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.cluster.service;
+
+import org.opensearch.common.AdjustableSemaphore;
+
+import java.util.Objects;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+
+/**
+ * Base class for Throttling logic.
+ * It provides throttling functionality over multiple keys.
+ * It provides the functionality of enable/disable throttling using enableThrottling variable.
+ *
+ * @param <T> the type of key on which we want to do throttling.
+ */
+public class Throttler<T> {
```

Comment:
This class never removes the semaphores, only creates new ones. I would suggest to introduce the reasonable bounds to the amount of the semaphores being kept (fe as constructor argument). 

Also, this class is never used (besides tests), why it is needed? It would be useful to see it is the context to justify why semaphore management has to be centralized (introducing additional contention).

Reply:
> This class never removes the semaphores, only creates new ones. I would suggest to introduce the reasonable bounds to the amount of the semaphores being kept (fe as constructor argument).

This class has `removeThrottlingLimit` method, which removes the semaphores from the map. 

> Also, this class is never used (besides tests), why it is needed? It would be useful to see it is the context to justify why semaphore management has to be centralized (introducing additional contention).

This is the first of multiple PRs planned for this SIM(https://github.com/opensearch-project/OpenSearch/issues/479#issuecomment-1142040484). In upcoming PRs we will use this Throttler class for throttling tasks at the master node.

Reply:
:+1: (missed the `removeThrottlingLimit`), thx